src/cgutils.cpp

```diff
@@ -1148,6 +1198,8 @@ static jl_cgval_t typed_load(Value *ptr, Value *idx_0based, jl_value_t *jltype,
     //else {
         Instruction *load = builder.CreateAlignedLoad(data, isboxed ?
             alignment : julia_alignment(data, jltype, alignment), false);
+        if (isboxed)
+            load = maybe_mark_load_dereferenceable(load, true, jltype);
```

Comment:
IIUC, this is marking the load from all pointers as being dereferencable (or null), but that would let LLVM hoist it above the bounds check (for example, the one in that PR).